### PR TITLE
nss: do not return IPv6 address when IPv6 is disabled

### LINKED ISSUE #23152: nss-myhostname returns IPv6 for localhost even with IPv6 disabled
**systemd version the issue has been seen with**
 > latest main

**Used distribution**
 > Fedora 35

This appears to be the same case as https://github.com/systemd/systemd/issues/18812#issuecomment-786700083:

With IPv6 on:
```
# sysctl -w  net.ipv6.conf.all.disable_ipv6=0
net.ipv6.conf.all.disable_ipv6 = 0
# getent ahosts localhost
::1             STREAM localhost
::1             DGRAM  
::1             RAW    
127.0.0.1       STREAM 
127.0.0.1       DGRAM  
127.0.0.1       RAW    
# getent -s resolve hosts localhost
::1             localhost
# getent -s myhostname hosts localhost
::1             localhost
```

With IPv6 off:
```
# sysctl -w  net.ipv6.conf.all.disable_ipv6=1
net.ipv6.conf.all.disable_ipv6 = 1
# getent ahosts localhost
127.0.0.1       STREAM localhost
127.0.0.1       DGRAM  
127.0.0.1       RAW    
# getent -s resolve hosts localhost
127.0.0.1       localhost
# getent -s myhostname hosts localhost
::1             localhost
```

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #23152.

### CODE REVIEW

**Comment by yuwata** [comment]
cc @mrc0mmand

**Review by bluca — APPROVED** [review]

**Comment by mrc0mmand** [comment]
Interesting, I rebased this on top of #23104 and now `nss-myhostname` doesn't return anything with IPv6 disabled:

```
# sysctl -w net.ipv6.conf.all.disable_ipv6=1
# getent ahosts localhost
127.0.0.1       STREAM localhost
127.0.0.1       DGRAM  
127.0.0.1       RAW    
# getent -s resolve hosts localhost
127.0.0.1       localhost
# getent -s myhostname hosts localhost
```

With IPv6 it works as before:
```
# getent ahosts localhost
::1             STREAM localhost
::1             DGRAM  
::1             RAW    
127.0.0.1       STREAM 
127.0.0.1       DGRAM  
127.0.0.1       RAW    
# getent -s resolve hosts localhost
::1             localhost
# getent -s myhostname hosts localhost
::1             localhost
```

Here's an strace of `getent -s myhostname hosts localhost` with IPv6 disabled for good measure:

```
# strace -yfs 500 getent -s myhostname hosts localhost
execve("/usr/bin/getent", ["getent", "-s", "myhostname", "hosts", "localhost"], 0x7fff368ede40 /* 18 vars */) = 0
brk(NULL)                               = 0x55c271cfc000
arch_prctl(0x3001 /* ARCH_??? */, 0x7fff61a17ad0) = -1 EINVAL (Invalid argument)
access("/etc/ld.so.preload", R_OK)      = -1 ENOENT (No such file or directory)
openat(AT_FDCWD</root>, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3</etc/ld.so.cache>
newfstatat(3</etc/ld.so.cache>, "", {st_mode=S_IFREG|0644, st_size=9187, ...}, AT_EMPTY_PATH) = 0
mmap(NULL, 9187, PROT_READ, MAP_PRIVATE, 3</etc/ld.so.cache>, 0) = 0x7f591bdb5000
close(3</etc/ld.so.cache>)              = 0
openat(AT_FDCWD</root>, "/lib64/libc.so.6", O_RDONLY|O_CLOEXEC) = 3</usr/lib64/libc.so.6>
read(3</usr/lib64/libc.so.6>, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\240\327\2\0\0\0\0\0@\0\0\0\0\0\0\0\250\341\37\0\0\0\0\0\0\0\0\0@\08\0\16\0@\0C\0B\0\6\0\0\0\4\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0\20\3\0\0\0\0\0\0\20\3\0\0\0\0\0\0\10\0\0\0\0\0\0\0\3\0\0\0\4\0\0\0 \206\34\0\0\0\0\0 \206\34\0\0\0\0\0 \206\34\0\0\0\0\0\34\0\0\0\0\0\0\0\34\0\0\0\0\0\0\0\20\0\0\0\0\0\0\0\1\0\0\0\4\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\230\275\2\0\0\0\0\0\230\275\2\0\0\0\0\0\0\20\0\0\0\0\0\0\1\0\0\0\5\0\0\0\0\300\2\0\0\0\0\0\0\300\2\0\0\0\0\0\0\300\2\0\0\0\0\0,S\27\0\0\0\0\0,S\27\0\0\0\0\0\0\20\0\0\0\0\0\0\1\0\0\0\4\0\0\0\0 \32\0\0\0\0\0\0 \32\0\0\0\0\0\0 \32\0\0\0\0\0\2548\5\0\0\0\0\0\2548\5\0\0\0\0\0\0\20\0\0\0\0\0\0\1\0\0\0\6\0\0\0\200e\37\0\0\0\0\0\200u\37\0\0\0\0\0\200u\37\0\0\0\0\0\220O\0\0\0\0\0\0000%\1\0\0\0\0\0\0\20\0\0\0\0\0\0\2\0\0\0\6\0\0\0\340\211\37\0\0\0\0\0\340\231\37\0\0\0\0\0\340\231\37\0\0\0\0\0\340\1\0\0\0\0\0\0\340\1\0\0\0\0\0\0\10\0\0\0\0\0\0\0\4\0\0\0\4\0\0\0P\3\0\0\0\0\0\0P\3\0\0\0\0\0\0P\3\0\0\0\0\0\0P\0\0\0\0\0\0\0P\0\0\0"..., 832) = 832
pread64(3</usr/lib64/libc.so.6>, "\6\0\0\0\4\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0\20\3\0\0\0\0\0\0\20\3\0\0\0\0\0\0\10\0\0\0\0\0\0\0\3\0\0\0\4\0\0\0 \206\34\0\0\0\0\0 \206\34\0\0\0\0\0 \206\34\0\0\0\0\0\34\0\0\0\0\0\0\0\34\0\0\0\0\0\0\0\20\0\0\0\0\0\0\0\1\0\0\0\4\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\230\275\2\0\0\0\0\0\230\275\2\0\0\0\0\0\0\20\0\0\0\0\0\0\1\0\0\0\5\0\0\0\0\300\2\0\0\0\0\0\0\300\2\0\0\0\0\0\0\300\2\0\0\0\0\0,S\27\0\0\0\0\0,S\27\0\0\0\0\0\0\20\0\0\0\0\0\0\1\0\0\0\4\0\0\0\0 \32\0\0\0\0\0\0 \32\0\0\0\0\0\0 \32\0\0\0\0\0\2548\5\0\0\0\0\0\2548\5\0\0\0\0\0\0\20\0\0\0\0\0\0\1\0\0\0\6\0\0\0\200e\37\0\0\0\0\0\200u\37\0\0\0\0\0\200u\37\0\0\0\0\0\220O\0\0\0\0\0\0000%\1\0\0\0\0\0\0\20\0\0\0\0\0\0\2\0\0\0\6\0\0\0\340\211\37\0\0\0\0\0\340\231\37\0\0\0\0\0\340\231\37\0\0\0\0\0\340\1\0\0\0\0\0\0\340\1\0\0\0\0\0\0\10\0\0\0\0\0\0\0\4\0\0\0\4\0\0\0P\3\0\0\0\0\0\0P\3\0\0\0\0\0\0P\3\0\0\0\0\0\0P\0\0\0\0\0\0\0P\0\0\0\0\0\0\0\10\0\0\0\0\0\0\0\4\0\0\0\4\0\0\0\240\3\0\0\0\0\0\0\240\3\0\0\0\0\0\0\240\3\0\0\0\0\0\0D\0\0\0\0\0\0\0D\0\0\0\0\0\0\0\4\0\0\0"..., 784, 64) = 784
pread64(3</usr/lib64/libc.so.6>, "\4\0\0\0@\0\0\0\5\0\0\0GNU\0\2\0\0\300\4\0\0\0\3\0\0\0\0\0\0\0\2\200\0\300\4\0\0\0\1\0\0\0\0\0\0\0\1\0\1\300\4\0\0\0;\10\0\0\0\0\0\0\2\0\1\300\4\0\0\0\17\0\0\0\0\0\0\0", 80, 848) = 80
pread64(3</usr/lib64/libc.so.6>, "\4\0\0\0\24\0\0\0\3\0\0\0GNU\0]c\3042{\246\315\321\350\271\236\261\300\251\360V\251\214g\341\4\0\0\0\20\0\0\0\1\0\0\0GNU\0\0\0\0\0\3\0\0\0\2\0\0\0\0\0\0\0", 68, 928) = 68
newfstatat(3</usr/lib64/libc.so.6>, "", {st_mode=S_IFREG|0755, st_size=2093672, ...}, AT_EMPTY_PATH) = 0
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f591bdb3000
pread64(3</usr/lib64/libc.so.6>, "\6\0\0\0\4\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0\20\3\0\0\0\0\0\0\20\3\0\0\0\0\0\0\10\0\0\0\0\0\0\0\3\0\0\0\4\0\0\0 \206\34\0\0\0\0\0 \206\34\0\0\0\0\0 \206\34\0\0\0\0\0\34\0\0\0\0\0\0\0\34\0\0\0\0\0\0\0\20\0\0\0\0\0\0\0\1\0\0\0\4\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\230\275\2\0\0\0\0\0\230\275\2\0\0\0\0\0\0\20\0\0\0\0\0\0\1\0\0\0\5\0\0\0\0\300\2\0\0\0\0\0\0\300\2\0\0\0\0\0\0\300\2\0\0\0\0\0,S\27\0\0\0\0\0,S\27\0\0\0\0\0\0\20\0\0\0\0\0\0\1\0\0\0\4\0\0\0\0 \32\0\0\0\0\0\0 \32\0\0\0\0\0\0 \32\0\0\0\0\0\2548\5\0\0\0\0\0\2548\5\0\0\0\0\0\0\20\0\0\0\0\0\0\1\0\0\0\6\0\0\0\200e\37\0\0\0\0\0\200u\37\0\0\0\0\0\200u\37\0\0\0\0\0\220O\0\0\0\0\0\0000%\1\0\0\0\0\0\0\20\0\0\0\0\0\0\2\0\0\0\6\0\0\0\340\211\37\0\0\0\0\0\340\231\37\0\0\0\0\0\340\231\37\0\0\0\0\0\340\1\0\0\0\0\0\0\340\1\0\0\0\0\0\0\10\0\0\0\0\0\0\0\4\0\0\0\4\0\0\0P\3\0\0\0\0\0\0P\3\0\0\0\0\0\0P\3\0\0\0\0\0\0P\0\0\0\0\0\0\0P\0\0\0\0\0\0\0\10\0\0\0\0\0\0\0\4\0\0\0\4\0\0\0\240\3\0\0\0\0\0\0\240\3\0\0\0\0\0\0\240\3\0\0\0\0\0\0D\0\0\0\0\0\0\0D\0\0\0\0\0\0\0\4\0\0\0"..., 784, 64) = 784
mmap(NULL, 2136752, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3</usr/lib64/libc.so.6>, 0) = 0x7f591bba9000
mprotect(0x7f591bbd5000, 1880064, PROT_NONE) = 0
mmap(0x7f591bbd5000, 1531904, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3</usr/lib64/libc.so.6>, 0x2c000) = 0x7f591bbd5000
mmap(0x7f591bd4b000, 344064, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3</usr/lib64/libc.so.6>, 0x1a2000) = 0x7f591bd4b000
mmap(0x7f591bda0000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3</usr/lib64/libc.so.6>, 0x1f6000) = 0x7f591bda0000
mmap(0x7f591bda6000, 51888, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f591bda6000
close(3</usr/lib64/libc.so.6>)          = 0
mmap(NULL, 12288, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f591bba6000
arch_prctl(ARCH_SET_FS, 0x7f591bba6740) = 0
set_tid_address(0x7f591bba6a10)         = 199
set_robust_list(0x7f591bba6a20, 24)     = 0
mprotect(0x7f591bda0000, 12288, PROT_READ) = 0
mprotect(0x55c27039e000, 4096, PROT_READ) = 0
mprotect(0x7f591bdec000, 8192, PROT_READ) = 0
prlimit64(0, RLIMIT_STACK, NULL, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
munmap(0x7f591bdb5000, 9187)            = 0
getrandom("\x43\x8a\xa2\xc9\x59\x79\x96\xb0", 8, GRND_NONBLOCK) = 8
brk(NULL)                               = 0x55c271cfc000
brk(0x55c271d1d000)                     = 0x55c271d1d000
openat(AT_FDCWD</root>, "/usr/lib/locale/locale-archive", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD</root>, "/usr/share/locale/locale.alias", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD</root>, "/usr/lib/locale/C.UTF-8/LC_IDENTIFICATION", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD</root>, "/usr/lib/locale/C.utf8/LC_IDENTIFICATION", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD</root>, "/usr/lib/locale/C/LC_IDENTIFICATION", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD</root>, "/etc/nsswitch.conf", {st_mode=S_IFREG|0644, st_size=2182, ...}, 0) = 0
newfstatat(AT_FDCWD</root>, "/", {st_mode=S_IFDIR|0755, st_size=1024, ...}, 0) = 0
openat(AT_FDCWD</root>, "/etc/nsswitch.conf", O_RDONLY|O_CLOEXEC) = 3</etc/nsswitch.conf>
newfstatat(3</etc/nsswitch.conf>, "", {st_mode=S_IFREG|0644, st_size=2182, ...}, AT_EMPTY_PATH) = 0
read(3</etc/nsswitch.conf>, "#\n# /etc/nsswitch.conf\n#\n# Name Service Switch config file. This file should be\n# sorted with the most-used services at the beginning.\n#\n# Valid databases are: aliases, ethers, group, gshadow, hosts,\n# initgroups, netgroup, networks, passwd, protocols, publickey,\n# rpc, services, and shadow.\n#\n# Valid service provider entries include (in alphabetical order):\n#\n#\tcompat\t\t\tUse /etc files plus *_compat pseudo-db\n#\tdb\t\t\tUse the pre-processed /var/db files\n#\tdns\t\t\tUse DNS (Domain Name Service)\n#\tfile"..., 1024) = 1024
read(3</etc/nsswitch.conf>, "ba winbind support\n#\twins\t\t\tUse Samba wins support\n#\twrapper\t\t\tUse wrapper module for testing\n#\n# Notes:\n#\n# 'sssd' performs its own 'files'-based caching, so it should generally\n# come before 'files'.\n#\n# WARNING: Running nscd with a secondary caching service like sssd may\n# \t   lead to unexpected behaviour, especially with how long\n# \t   entries are cached.\n#\n# Installation instructions:\n#\n# To use 'db', install the appropriate package(s) (provide 'makedb' and\n# libnss_db.so.*), and place the "..., 1024) = 1024
read(3</etc/nsswitch.conf>, " to default to the setting for group.\n# initgroups: files\nnetworks:   files dns\nprotocols:  files\npublickey:  files\nrpc:        files\n", 1024) = 134
read(3</etc/nsswitch.conf>, "", 1024)   = 0
newfstatat(3</etc/nsswitch.conf>, "", {st_mode=S_IFREG|0644, st_size=2182, ...}, AT_EMPTY_PATH) = 0
close(3</etc/nsswitch.conf>)            = 0
newfstatat(AT_FDCWD</root>, "/etc/resolv.conf", {st_mode=S_IFREG|0644, st_size=920, ...}, 0) = 0
openat(AT_FDCWD</root>, "/etc/host.conf", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
futex(0x7f591bdace0c, FUTEX_WAKE_PRIVATE, 2147483647) = 0
openat(AT_FDCWD</root>, "/etc/resolv.conf", O_RDONLY|O_CLOEXEC) = 3</run/systemd/resolve/stub-resolv.conf>
newfstatat(3</run/systemd/resolve/stub-resolv.conf>, "", {st_mode=S_IFREG|0644, st_size=920, ...}, AT_EMPTY_PATH) = 0
read(3</run/systemd/resolve/stub-resolv.conf>, "# This is /run/systemd/resolve/stub-resolv.conf managed by man:systemd-resolved(8).\n# Do not edit.\n#\n# This file might be symlinked as /etc/resolv.conf. If you're looking at\n# /etc/resolv.conf and seeing this text, you have followed the symlink.\n#\n# This is a dynamic resolv.conf file for connecting local clients to the\n# internal DNS stub resolver of systemd-resolved. This file lists all\n# configured search domains.\n#\n# Run \"resolvectl status\" to see details about the uplink DNS servers\n# curren"..., 4096) = 920
read(3</run/systemd/resolve/stub-resolv.conf>, "", 4096) = 0
newfstatat(3</run/systemd/resolve/stub-resolv.conf>, "", {st_mode=S_IFREG|0644, st_size=920, ...}, AT_EMPTY_PATH) = 0
close(3</run/systemd/resolve/stub-resolv.conf>) = 0
openat(AT_FDCWD</root>, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3</etc/ld.so.cache>
newfstatat(3</etc/ld.so.cache>, "", {st_mode=S_IFREG|0644, st_size=9187, ...}, AT_EMPTY_PATH) = 0
mmap(NULL, 9187, PROT_READ, MAP_PRIVATE, 3</etc/ld.so.cache>, 0) = 0x7f591bdb5000
close(3</etc/ld.so.cache>)              = 0
openat(AT_FDCWD</root>, "/lib64/libnss_myhostname.so.2", O_RDONLY|O_CLOEXEC) = 3</usr/lib64/libnss_myhostname.so.2>
read(3</usr/lib64/libnss_myhostname.so.2>, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\340\1\1\0\0\0\0\0@\0\0\0\0\0\0\08\305\17\0\0\0\0\0\0\0\0\0@\08\0\f\0@\0\35\0\34\0\1\0\0\0\4\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\20\353\0\0\0\0\0\0\20\353\0\0\0\0\0\0\0\20\0\0\0\0\0\0\1\0\0\0\5\0\0\0\0\360\0\0\0\0\0\0\0\360\0\0\0\0\0\0\0\360\0\0\0\0\0\0\375S\n\0\0\0\0\0\375S\n\0\0\0\0\0\0\20\0\0\0\0\0\0\1\0\0\0\4\0\0\0\0P\v\0\0\0\0\0\0P\v\0\0\0\0\0\0P\v\0\0\0\0\0\330\222\3\0\0\0\0\0\330\222\3\0\0\0\0\0\0\20\0\0\0\0\0\0\1\0\0\0\6\0\0\0 \351\16\0\0\0\0\0 \371\16\0\0\0\0\0 \371\16\0\0\0\0\0\210\332\0\0\0\0\0\0x\343\0\0\0\0\0\0\0\20\0\0\0\0\0\0\2\0\0\0\6\0\0\0\200\263\17\0\0\0\0\0\200\303\17\0\0\0\0\0\200\303\17\0\0\0\0\0\0\2\0\0\0\0\0\0\0\2\0\0\0\0\0\0\10\0\0\0\0\0\0\0\4\0\0\0\4\0\0\0\340\2\0\0\0\0\0\0\340\2\0\0\0\0\0\0\340\2\0\0\0\0\0\0000\0\0\0\0\0\0\0000\0\0\0\0\0\0\0\10\0\0\0\0\0\0\0\4\0\0\0\4\0\0\0\20\3\0\0\0\0\0\0\20\3\0\0\0\0\0\0\20\3\0\0\0\0\0\0$\0\0\0\0\0\0\0$\0\0\0\0\0\0\0\4\0\0\0\0\0\0\0\7\0\0\0\4\0\0\0 \351\16\0\0\0\0\0 \371\16\0\0\0\0\0 \371\16\0\0\0\0\0000\0\0\0\0\0\0\0\220\0\0\0"..., 832) = 832
newfstatat(3</usr/lib64/libnss_myhostname.so.2>, "", {st_mode=S_IFREG|0755, st_size=1035384, ...}, AT_EMPTY_PATH) = 0
mmap(NULL, 1039512, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3</usr/lib64/libnss_myhostname.so.2>, 0) = 0x7f591baa8000
mmap(0x7f591bab7000, 679936, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3</usr/lib64/libnss_myhostname.so.2>, 0xf000) = 0x7f591bab7000
mmap(0x7f591bb5d000, 237568, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3</usr/lib64/libnss_myhostname.so.2>, 0xb5000) = 0x7f591bb5d000
mmap(0x7f591bb97000, 61440, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3</usr/lib64/libnss_myhostname.so.2>, 0xee000) = 0x7f591bb97000
close(3</usr/lib64/libnss_myhostname.so.2>) = 0
mprotect(0x7f591bb97000, 57344, PROT_READ) = 0
munmap(0x7f591bdb5000, 9187)            = 0
rt_sigprocmask(SIG_BLOCK, [HUP USR1 USR2 PIPE ALRM CHLD TSTP URG VTALRM PROF WINCH IO], [], 8) = 0
access("/proc/net/if_inet6", F_OK)      = 0
openat(AT_FDCWD</root>, "/proc/sys/net/ipv6/conf/all/disable_ipv6", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 3</proc/sys/net/ipv6/conf/all/disable_ipv6>
newfstatat(3</proc/sys/net/ipv6/conf/all/disable_ipv6>, "", {st_mode=S_IFREG|0644, st_size=0, ...}, AT_EMPTY_PATH) = 0
read(3</proc/sys/net/ipv6/conf/all/disable_ipv6>, "1\n", 4104) = 2
close(3</proc/sys/net/ipv6/conf/all/disable_ipv6>) = 0
openat(AT_FDCWD</root>, "/proc/sys/net/ipv6/conf/all/disable_ipv6", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 3</proc/sys/net/ipv6/conf/all/disable_ipv6>
newfstatat(3</proc/sys/net/ipv6/conf/all/disable_ipv6>, "", {st_mode=S_IFREG|0644, st_size=0, ...}, AT_EMPTY_PATH) = 0
read(3</proc/sys/net/ipv6/conf/all/disable_ipv6>, "1\n", 4104) = 2
close(3</proc/sys/net/ipv6/conf/all/disable_ipv6>) = 0
openat(AT_FDCWD</root>, "/proc/sys/net/ipv6/conf/all/disable_ipv6", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 3</proc/sys/net/ipv6/conf/all/disable_ipv6>
newfstatat(3</proc/sys/net/ipv6/conf/all/disable_ipv6>, "", {st_mode=S_IFREG|0644, st_size=0, ...}, AT_EMPTY_PATH) = 0
read(3</proc/sys/net/ipv6/conf/all/disable_ipv6>, "1\n", 4104) = 2
close(3</proc/sys/net/ipv6/conf/all/disable_ipv6>) = 0
openat(AT_FDCWD</root>, "/proc/sys/net/ipv6/conf/all/disable_ipv6", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 3</proc/sys/net/ipv6/conf/all/disable_ipv6>
newfstatat(3</proc/sys/net/ipv6/conf/all/disable_ipv6>, "", {st_mode=S_IFREG|0644, st_size=0, ...}, AT_EMPTY_PATH) = 0
read(3</proc/sys/net/ipv6/conf/all/disable_ipv6>, "1\n", 4104) = 2
close(3</proc/sys/net/ipv6/conf/all/disable_ipv6>) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
exit_group(0)                           = ?
+++ exited with 0 +++

```

File: src/resolve/resolved-etc-hosts.c [hunk 1]
```diff
@@ -440,7 +440,7 @@ int manager_etc_hosts_lookup(Manager *m, DnsQuestion* q, DnsAnswer **answer) {
 
                 if (IN_SET(t->type, DNS_TYPE_A, DNS_TYPE_ANY))
                         found_a = true;
-                if (IN_SET(t->type, DNS_TYPE_AAAA, DNS_TYPE_ANY))
+                if (IN_SET(t->type, DNS_TYPE_AAAA, DNS_TYPE_ANY) && socket_ipv6_is_enabled())
```
Comment (poettering) [inline src/resolve/resolved-etc-hosts.c]: i am not convinced we should do this. if we synthesize localhost, then yes, we should suppress the AAAA record if ipv6 is off. But I am pretty sure we should not second guess /etc/hosts. People should not list ::1 there if they don#t want it resolved.

Quite frnakly, they should drop any entry for "localhost" altogether, and let us synthesize it.

**Comment by poettering** [comment]
patch looks goot where it touches our synthesis of recrods, but I am pretty sure we should leave /etc/hosts alone.

**Comment by yuwata** [comment]
OK. The change for resolved is dropped. PTAL.
Note, `getent -s files hosts localhost` also returns `::1` even if IPv6 is disabled.

**Comment by mrc0mmand** [comment]
So, in the systemd test images `/etc/hosts` appears to be empty by default. So with this patch set I still get no IPv4 when IPv6 is disabled, even, if I understand the discussion here correctly, it should be synthesized in that case:

`/etc/hosts`:
```
# cat /etc/hosts
1.2.3.1 ns1.signed.test
```

IPv6 enabled:
```
# getent ahosts localhost
::1             STREAM localhost
::1             DGRAM  
::1             RAW    
127.0.0.1       STREAM 
127.0.0.1       DGRAM  
127.0.0.1       RAW    
# getent -s resolve hosts localhost
::1             localhost
# getent -s myhostname hosts localhost
::1             localhost
```

IPv6 disabled:
```
# sysctl -w net.ipv6.conf.all.disable_ipv6=1
# getent ahosts localhost
127.0.0.1       STREAM localhost
127.0.0.1       DGRAM  
127.0.0.1       RAW  
# getent -s resolve hosts localhost
127.0.0.1       localhost
# getent -s myhostname hosts localhost
```

I admit that I'm a bit confused here, so the main question is - what should be the expected state?

**Comment by yuwata** [comment]
@mrc0mmand I overlooked that myhostname does not provide IPv4 address. Will take a look.

**Comment by yuwata** [comment]
@mrc0mmand Could you take a look again? I tried to cherry-pick #23104 and re-enable the commented out lines, and the test succeeds. But please double check.

**Comment by mrc0mmand** [comment]
> @mrc0mmand Could you take a look again? I tried to cherry-pick #23104 and re-enable the commented out lines, and the test succeeds. But please double check.

It indeed succeeds, thank you!

**Pull request merged**